fix: Fix `TableTextRetriever` for input consisting of tables only

jackapbutler opened the pull request:
### Related Issues
- fixes #4013 

### Proposed Changes:
* The `language_model3` within `TriAdaptiveModel` expects kwargs which don't contain the "passage_" prefix. I changed these to conform with the function kwargs.
* The `TableTextRetriever` is composed of an underlying `BertModel` which in turn contains a `BertEmbedding` layer which expects a 2D batch shape of [batch_size, max_seq_len] but the `TableTextRetriever` performed inference on a 3D tensor with shape [batch_size, number_of_hard_negative_passages, max_seq_len]. I changed this to conform with the 2D expectation by adding this change to `retriever. _get_predictions()` 

### How did you test it?
I tested it locally using the code snippet provided in the issue.

### Checklist
- [X] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [X] I have updated the related issue with new insights and changes
- [X] I added tests that demonstrate the correct behavior of the change
- [X] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- [X] I documented my code
- [X] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

bogdankostic commented on haystack/modeling/model/triadaptive_model.py:
```diff
@@ -310,9 +310,9 @@ def forward_lm(self, **kwargs):
             if all(table_mask):
                 pooled_output2, _ = self.language_model3(
-                    passage_input_ids=kwargs["passage_input_ids"],
-                    passage_segment_ids=kwargs["table_segment_ids"],
-                    passage_attention_mask=kwargs["passage_attention_mask"],
+                    input_ids=kwargs["passage_input_ids"],
+                    segment_ids=kwargs["table_segment_ids"],
+                    attention_mask=kwargs["passage_attention_mask"],
                     output_hidden_states=False,
                     output_attentions=False,
                 )
```
Let's make the input two-dimensional here, similar to the case where we only have text Documents.
```suggestion
            if all(table_mask):
                # Make input two-dimensional
                max_seq_len = kwargs["table_input_ids"].shape[-1]
                input_ids = kwargs["table_input_ids"].view(-1, max_seq_len)
                attention_mask = kwargs["table_attention_mask"].view(-1, max_seq_len)
                segment_ids = kwargs["table_segment_ids"].view(-1, max_seq_len)

                pooled_output2, _ = self.language_model3(
                    input_ids=input_ids,
                    segment_ids=segment_ids,
                    attention_mask=attention_mask,
                    output_hidden_states=False,
                    output_attentions=False,
                )
```

bogdankostic reviewed (changes requested):
Thanks for creating this PR @jackapbutler! I added a few comments on how we can improve it, let me know what you think of them.

jackapbutler commented:
Thanks @bogdankostic for your feedback, I've just updated the pull request now and added reorganised the tests now 🙂. 

I noticed the `TriAdaptiveModel` had duplicate executions of the following lines within each clause of the `if "passage_input_ids" in kwargs.keys()` condition branch.

```python
# Make input two-dimensional
max_seq_len = kwargs["passage_input_ids"].shape[-1]
passage_input_ids = kwargs["passage_input_ids"].view(-1, max_seq_len)
passage_attention_mask = kwargs["passage_attention_mask"].view(-1, max_seq_len)
passage_segment_ids = kwargs["passage_segment_ids"].view(-1, max_seq_len)
```

I've therefore placed these at the top outside of any further conditions (`if all(table_mask)`, `elif any(table_mask)`, else) so to remove the duplication, is that correct?

bogdankostic commented on haystack/nodes/retriever/dense.py:
```diff
@@ -1125,8 +1125,6 @@ def _get_predictions(self, dicts: List[Dict[str, Any]]) -> Dict[str, np.ndarray]
             disable=disable_tqdm,
         ) as progress_bar:
             for batch in data_loader:
-                batch = {key: batch[key].to(self.devices[0]) for key in batch}
```
We need to keep this line in order to use the GPU when doing inference.

jackapbutler replied:
Thanks for catching that, sorry @bogdankostic 🤦‍♂️!

bogdankostic commented on test/nodes/test_retriever.py:
```diff
@@ -444,6 +444,36 @@ def test_table_text_retriever_embedding(document_store, retriever, docs):
+@pytest.mark.parametrize("retriever", ["table_text_retriever"], indirect=True)
+@pytest.mark.parametrize("document_store", ["memory"], indirect=True)
+@pytest.mark.embedding_dim(512)
+def test_table_text_retriever_embedding_only_table(document_store, retriever):
```
This should be renamed:
```suggestion
def test_table_text_retriever_embedding_only_text(document_store, retriever):
```

bogdankostic reviewed (changes requested):
Thank you for removing the duplicate code, makes totally sense! I added two minor comments, the PR should be good to go after resolving these :)

bogdankostic commented on test/nodes/test_retriever.py:
```diff
@@ -444,6 +444,36 @@ def test_table_text_retriever_embedding(document_store, retriever, docs):
+        Document(content="This is a test", content_type="text"),
+        Document(content="This is another test", content_type="text"),
+    ]
+    retriever = retriever(document_store=document_store)
```
This line needs to be removed. (The retriever is already initialized via the fixture.)

bogdankostic commented on test/nodes/test_retriever.py:
```diff
@@ -444,6 +444,36 @@ def test_table_text_retriever_embedding(document_store, retriever, docs):
+        content=pd.DataFrame(columns=["id", "text"], data=[["1", "This is a test"], ["2", "This is another test"]]),
+        content_type="table",
+    )
+    retriever = retriever(document_store=document_store)
```
This line needs to be removed as well.

jackapbutler commented:
Hey @bogdankostic, I've just noticed the open PR that mentions this pull request. Will I pause on this, rebase once that is merged into main and trigger CI again?

bogdankostic reviewed (approved):
Looking good to me now, thank you for your contribution! :)
